cuda 버전 추가 / zoom_in_out 기능 추가

creeras opened the pull request:
cuda 버전 추가되었습니다. 
cuda_zoomin 기능 추가되었습니다. 재실행 없이 화면에서 그대로 변경이 적용됩니다.

creeras committed:
Add files via upload

CUDA 명령어(cupy)로 작동하는 버전입니다.

creeras committed:
Update and rename ptf_zoom_gpu.py to ptf_zoom_cuda.py

파일명 변경합니다. 
gpu -> cuda

creeras committed:
Create README.md

creeras committed:
Add files via upload

creeras committed:
Update README.md

creeras committed:
Update README.md

creeras committed:
Update README.md

creeras committed:
Update ptf2_cuda.py

memory 관리 개선

creeras committed:
Update README.md

creeras committed:
테트레이션 이미지에서 시작/끝점을 클릭해 바로 줌인할 수 있게 함.

creeras committed:
Update ptf2_cuda_zoomin.py

더블클릭시 무시. 줌인사각형 두께=2

creeras committed:
Update README.md

creeras committed:
Update README.md

줄바꿈 수정.

creeras committed:
Update README.md

보기좋게 수정함.

creeras committed:
Create ptf2_cuda_zoomin_structure.md

Markdown Structure

creeras committed:
Update README.md

creeras committed:
Update and rename ptf2_cuda_zoomin.py to ptf2_cuda_zoom.py

zoomout 기능(버튼) 추가

creeras committed:
Delete cuda/ptf2_cuda_zoomin_structure.md

불필요

creeras committed:
Update README.md

### ptf2_cuda_zoom.py  
관련 설명 수정.

creeras reviewed (commented):
ptf2_cuda_zoom.py 에 zoom out 기능이 추가되었습니다.

sasak2 commented:
colab에서 런타임 T4 GPU로 실행해보니 n=500기준 2~3초만에 나오네요 :+1: :+1: 
선생님 너무 좋습니다!! 따로 코랩에 이식해서 내면 많은 사람들에게 접근성도 좋고 속도도 빠르게 할 수 있을 것 같아요!!

chitenshi commented:
colab에서 돌려봤는데 정말 빠릅니다
compute_tetration_divergence함수에서 반복문 부분을
```
    for k i tqdm(range(max_iter), desc="Calculating Tetration", ncols=100):
        cp.power(c, z, out=z)
    cp.logical_or(cp.isinf(z), cp.isnan(z), out=divergence_map)
    mask = cp.abs(z) > escape_radius
    divergence_map[mask] = True
```
이렇게 수정하면 좀 더 빠르게 작동하지 않을까 생각합니다
이 코드를 당장 colab에서 테스트해보고 싶었지만 할당량이 넘어서 미리 실험해보지 못했습니다.

creeras committed:
Update README.md

creeras committed:
Update README.md

creeras committed:
Update ptf2_cuda_zoom.py

창 프레임 구조 변경. 
정밀도 설정 변경. 
프로그레스 바 설정 등.

creeras reviewed (commented):
1. ptf2_cuda_zoom.py 의 창 프레임 구조를 변경하였습니다. 
![스크린샷 2024-05-27 105904](https://github.com/DMTPARK/mytetration/assets/4322821/28c6260b-0e51-480b-88b8-9999d5199126)

2. 복소수 정밀도를 (eps값에 따라) 유동적으로 적용하게 하였습니다. 
eps>=1e-4 일 때는 cp.complex64 를 배열에 할당하여 계산 속도를 높였습니다. 
(이 때는 계산 속도가 5배 정도 빨라집니다)
그렇지 않으면 cp.complex128(=float64에 해당) 이 적용되고, 원래의 정밀도를 갖습니다. 
**[README.md 에 나와 있는 확대 한계](https://github.com/DMTPARK/mytetration/blob/main/README.md#ptf%EC%BD%94%EB%93%9C%EC%97%90-%EB%8C%80%ED%95%9C-%EC%BD%94%EB%A9%98%ED%8A%B8)** 와 관련이 있습니다. 

더 이상의 정밀한 계산(=더 깊은 zoom in)을 하려면 임의 정밀도 라이브러리(mpmath) 등을 이용해야 한다고 하는데, 
(해보진 않았지만) 아마도 계산 속도에 상당한 영향을 줄 것으로 생각됩니다.

creeras committed:
Update ptf2_cuda_zoom.py

1. Zoom Out 버튼에 Invert 옵션으로 Zoom In 기능 추가.
2. 전체 White인 cmap 이 전체 black 으로 그려지는 버그 수정.

creeras committed:
Merge branch 'DMTPARK:main' into main

creeras committed:
Update ptf2_cuda_zoom.py

Save Image 기능 추가 : .py 파일의 아래 날자폴더에 저장.

creeras reviewed (commented):
1. Zoom Out 버튼이 Zoom Out 버튼으로도 작동하게 하였습니다. 
2. 이미지 저장 기능을 추가하였습니다. 
.\오늘날자\mytetration_x-0.19271201160326912_y0.23536223034290904_eps0.0005634972685472084.png 
형식으로 저장됩니다.

creeras committed:
Update README.md

colab 버전 링크 추가
